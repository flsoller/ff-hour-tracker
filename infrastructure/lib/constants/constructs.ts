export const USERS_SERVICE = {
  NAME: "hour-tracker-users-service",
  LAMBDA_INTEGRATION: "hour-tracker-users-service-lambda-integration",
};

export const API_GATEWAY = {
  NAME: "hour-tracker-api-gateway",
};

export const DEFAULT_AUTHORIZER = {
  NAME: "hour-tracker-default-authorizer-service",
  LAMBDA_AUTHORIZER: "hour-tracker-default-lambda-authorizer",
};

export const AUTHENTICATOR_SERVICE = {
  NAME: "hour-tracker-authentication-service",
  LAMBDA_INTEGRATION: "hour-tracker-authentication-service-lambda-integration",
};

export const ORGANIZATION_MANAGER = {
  NAME: "hour-tracker-organization-manager",
};
