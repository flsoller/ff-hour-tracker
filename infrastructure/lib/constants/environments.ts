export const ENVIRONMENT_PARAMS = {
  DATABASE_URL_KEY: "HOUR_TRACKER_DATABASE_URL",
  JWT_SECRET: "HOUR_TRACKER_JWT_SECRET",
  API_DOMAIN_NAME: "HOUR_TRACKER_API_DOMAIN_NAME",
  API_DOMAIN_CERT_ARN: "HOUR_TRACKER_API_DOMAIN_CERT_ARN",
};
