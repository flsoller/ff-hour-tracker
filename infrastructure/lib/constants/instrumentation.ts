export const DEFAULT_INSTRUMENTATION_CONFIG = {
  NEW_RELIC_EXTENSION_SEND_FUNCTION_LOGS: "true",
  NEW_RELIC_LAMBDA_EXTENSION_ENABLED: "true",
  NEW_RELIC_DATA_COLLECTION_TIMEOUT: "10s",
  NEW_RELIC_IGNORE_EXTENSION_CHECKS: "all",
  NEW_RELIC_NO_CONFIG_FILE: "true",
};
