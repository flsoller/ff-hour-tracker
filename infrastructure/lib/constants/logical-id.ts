const AppName = "HourTrackerAPI";

export const LOGICAL_ID = {
  HOUR_TRACKER_API_GATEWAY: `${AppName}-Gateway`,
  HOUR_TRACKER_API_AUTHENTICATOR: `${AppName}-Authenticator`,
  HOUR_TRACKER_API_AUTHORIZER: `${AppName}-Authorizer`,
  HOUR_TRACKER_API_MEMBERS: `${AppName}-Members`,
  HOUR_TRACKER_API_ORG_MANAGER: `${AppName}-OrgManager`,
};
