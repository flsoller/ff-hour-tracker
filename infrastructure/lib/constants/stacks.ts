export const HOUR_TRACKER = {
  API_AUTHORIZER: "hour-tracker-api-authorizer",
  API_AUTHENTICATOR: "hour-tracker-api-authenticator",
  API_GATEWAY: "hour-tracker-api-gatewayv2",
  API_MEMBERS: "hour-tracker-api-members",
};
