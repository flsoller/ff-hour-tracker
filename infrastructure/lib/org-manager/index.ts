import * as cdk from "aws-cdk-lib";
import * as ecr from "aws-cdk-lib/aws-ecr";
import * as ssm from "aws-cdk-lib/aws-ssm";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { Duration } from "aws-cdk-lib";
import { Construct } from "constructs";
import { RemovalPolicy } from "aws-cdk-lib";
import { HOUR_TRACKER_ECR_REPO_NAMES } from "../constants/ecr";
import { ORGANIZATION_MANAGER } from "../constants/constructs";
import { ENVIRONMENT_PARAMS } from "../constants/environments";
import { LOGICAL_ID } from "../constants/logical-id";

export class HourTrackerOrganizationManager extends cdk.Stack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const dbConnectionString = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.DATABASE_URL_KEY
    );
    const supportUserEmail = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.SUPPORT_USER_EMAIL
    );
    const supportUserPassword = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.SUPPORT_USER_PW
    );

    const newRelicAccountId = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.NEW_RELIC_ACCOUNT_ID
    );

    const newRelicIngestLicense = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.NEW_RELIC_LICENSE_KEY
    );

    const hashOrVersion = new cdk.CfnParameter(this, "hashOrVersion", {
      description: "The version or hash used to reference the images",
      type: "String",
    }).valueAsString;

    const ecrRepo = ecr.Repository.fromRepositoryName(
      this,
      `${id}-repo`,
      HOUR_TRACKER_ECR_REPO_NAMES.ORGANIZATION_MANAGER
    );

    const organizationManager = new lambda.DockerImageFunction(
      this,
      ORGANIZATION_MANAGER.NAME,
      {
        code: lambda.DockerImageCode.fromEcr(ecrRepo, {
          tagOrDigest: hashOrVersion,
        }),
        architecture: lambda.Architecture.ARM_64,
        environment: {
          DATABASE_URL: dbConnectionString,
          SUPPORT_USER_EMAIL: supportUserEmail,
          SUPPORT_USER_PW: supportUserPassword,
          NEW_RELIC_LAMBDA_HANDLER: "apps/org-manager/index.handler",
          NEW_RELIC_ACCOUNT_ID: newRelicAccountId,
          NEW_RELIC_LICENSE_KEY: newRelicIngestLicense,
        },
        timeout: Duration.seconds(20),
        functionName: LOGICAL_ID.HOUR_TRACKER_API_ORG_MANAGER,
      }
    );
    organizationManager.applyRemovalPolicy(RemovalPolicy.DESTROY);
  }
}
