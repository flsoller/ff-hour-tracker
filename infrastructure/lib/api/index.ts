import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { AuthorizerService } from "./authorizer";
import { HourTrackerApiGateway } from "./gateway";
import { HOUR_TRACKER } from "../constants/stacks";
import * as ssm from "aws-cdk-lib/aws-ssm";
import { ENVIRONMENT_PARAMS } from "../constants/environments";
import { AuthenticationService } from "./authenticator";
import { MembersService } from "./members";

export class HourTrackerApi extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const hashOrVersion = new cdk.CfnParameter(this, "hashOrVersion", {
      description: "The version or hash used to reference the images",
      type: "String",
    }).valueAsString;

    const dbConnectionString = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.DATABASE_URL_KEY
    );

    const jwtSecretString = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.JWT_SECRET
    );

    const newRelicAccountId = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.NEW_RELIC_ACCOUNT_ID
    );

    const newRelicIngestLicense = ssm.StringParameter.valueForStringParameter(
      this,
      ENVIRONMENT_PARAMS.NEW_RELIC_LICENSE_KEY
    );

    const defaultAuthorizer = new AuthorizerService(
      this,
      HOUR_TRACKER.API_AUTHORIZER,
      {
        hashOrVersion,
        dbConnectionString,
        jwtSecretString,
        newRelicAccountId,
        newRelicIngestLicense,
      }
    );

    const apiGateway = new HourTrackerApiGateway(
      this,
      HOUR_TRACKER.API_GATEWAY,
      {
        authService: defaultAuthorizer.authorizer,
      }
    );

    new AuthenticationService(this, HOUR_TRACKER.API_AUTHENTICATOR, {
      apiGateway: apiGateway.httpApiGateway,
      dbConnectionString,
      hashOrVersion,
      jwtSecretString,
      newRelicAccountId,
      newRelicIngestLicense,
    });

    new MembersService(this, HOUR_TRACKER.API_MEMBERS, {
      apiGateway: apiGateway.httpApiGateway,
      dbConnectionString,
      hashOrVersion,
      newRelicAccountId,
      newRelicIngestLicense,
    });
  }
}
