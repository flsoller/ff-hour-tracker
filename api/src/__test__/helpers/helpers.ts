import { prisma } from '../../utils/prisma';
import { registerUser } from '../../services/auth';
import { IRegisterUser } from '@hour-tracker/core-types/api/auth';
import supertest from 'supertest';
import { Role } from '@prisma/client';

async function createOrganizations() {
  const organization = await prisma.organization.create({
    data: {
      name: 'organization',
      description: 'organization description',
    },
  });

  const otherOrganization = await prisma.organization.create({
    data: {
      name: 'otherOrganization',
      description: 'otherOrganization description',
    },
  });

  return [organization, otherOrganization];
}

async function createMemberForOrganization(
  orgId: string,
  emailAddress = 'test@mail.com',
  firstName = 'Test',
  lastName = 'User',
) {
  return prisma.member.create({
    data: {
      firstName,
      lastName,
      emailAddress,
      orgId,
    },
  });
}

async function createUserForOrganization(orgId: string, options = {}) {
  const params: IRegisterUser = {
    emailAddress: 'testuser@db.com',
    password: '12345',
    name: 'Mr. Test',
    orgId,
    role: Role.USER,
    ...options,
  };

  return registerUser(params);
}

async function loginTestUser(
  app: Express.Application,
  emailAddress: string,
): Promise<string> {
  return supertest(app)
    .post('/api/v0/auth/signin')
    .send({
      emailAddress,
      password: '12345',
    })
    .then((res) => res.body.accessToken);
}

export {
  createOrganizations,
  createMemberForOrganization,
  createUserForOrganization,
  loginTestUser,
};
