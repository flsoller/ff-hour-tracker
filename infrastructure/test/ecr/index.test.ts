import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { HOUR_TRACKER_ECR_REPO_NAMES } from "../../lib/constants/ecr";
import { HourTrackerImageRepositories } from "../../lib/ecr";
import { cdkResourceFinder } from "../helpers";

let app: cdk.App;
let ecrStack: HourTrackerImageRepositories;
let template: Template;

beforeEach(() => {
  app = new cdk.App({
    context: { "@aws-cdk/core:newStyleStackSynthesis": false },
  });
  ecrStack = new HourTrackerImageRepositories(
    app,
    "HourTrackerImageRepository",
  );
  template = Template.fromStack(ecrStack);
});

describe("HourTrackerImageRepositories", () => {
  it("should contain the correct amount of repositories", () => {
    template.resourceCountIs("AWS::ECR::Repository", 4);
  });

  describe("API_AUTHORIZER", () => {
    it("should have the correct repository parameters", () => {
      const resource = cdkResourceFinder(
        template,
        HOUR_TRACKER_ECR_REPO_NAMES.API_AUTHORIZER,
      );
      expect(resource).toMatchSnapshot();
    });
  });

  describe("API_AUTHENTICATOR", () => {
    it("should have the correct repository parameters", () => {
      const resource = cdkResourceFinder(
        template,
        HOUR_TRACKER_ECR_REPO_NAMES.API_AUTHENTICATOR,
      );
      expect(resource).toMatchSnapshot();
    });
  });

  describe("ORGANIZATION_MANAGER", () => {
    it("should have the correct repository parameters", () => {
      const resource = cdkResourceFinder(
        template,
        HOUR_TRACKER_ECR_REPO_NAMES.ORGANIZATION_MANAGER,
      );
      expect(resource).toMatchSnapshot();
    });
  });

  describe("MEMBERS_SERVICE", () => {
    it("should have the correct repository parameters", () => {
      const resource = cdkResourceFinder(
        template,
        HOUR_TRACKER_ECR_REPO_NAMES.API_MEMBERS,
      );
      expect(resource).toMatchSnapshot();
    });
  });
});
