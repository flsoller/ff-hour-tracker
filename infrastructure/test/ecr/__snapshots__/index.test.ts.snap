// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HourTrackerImageRepositories API_AUTHENTICATOR should have the correct repository parameters 1`] = `
{
  "DeletionPolicy": "Delete",
  "Properties": {
    "EmptyOnDelete": true,
    "ImageScanningConfiguration": {
      "ScanOnPush": false,
    },
    "LifecyclePolicy": {
      "LifecyclePolicyText": "{"rules":[{"rulePriority":1,"description":"Keep max 2 versioned images identified by 'vX.X.X' pattern","selection":{"tagStatus":"tagged","tagPatternList":["v*.*.*"],"countType":"imageCountMoreThan","countNumber":2},"action":{"type":"expire"}},{"rulePriority":2,"description":"Keep max 1 image that is not a tagged version","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1},"action":{"type":"expire"}}]}",
    },
    "RepositoryName": "hour-tracker-api-authenticator",
  },
  "Type": "AWS::ECR::Repository",
  "UpdateReplacePolicy": "Delete",
}
`;

exports[`HourTrackerImageRepositories API_AUTHORIZER should have the correct repository parameters 1`] = `
{
  "DeletionPolicy": "Delete",
  "Properties": {
    "EmptyOnDelete": true,
    "ImageScanningConfiguration": {
      "ScanOnPush": false,
    },
    "LifecyclePolicy": {
      "LifecyclePolicyText": "{"rules":[{"rulePriority":1,"description":"Keep max 2 versioned images identified by 'vX.X.X' pattern","selection":{"tagStatus":"tagged","tagPatternList":["v*.*.*"],"countType":"imageCountMoreThan","countNumber":2},"action":{"type":"expire"}},{"rulePriority":2,"description":"Keep max 1 image that is not a tagged version","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1},"action":{"type":"expire"}}]}",
    },
    "RepositoryName": "hour-tracker-api-authorizer",
  },
  "Type": "AWS::ECR::Repository",
  "UpdateReplacePolicy": "Delete",
}
`;

exports[`HourTrackerImageRepositories MEMBERS_SERVICE should have the correct repository parameters 1`] = `
{
  "DeletionPolicy": "Delete",
  "Properties": {
    "EmptyOnDelete": true,
    "ImageScanningConfiguration": {
      "ScanOnPush": false,
    },
    "LifecyclePolicy": {
      "LifecyclePolicyText": "{"rules":[{"rulePriority":1,"description":"Keep max 2 versioned images identified by 'vX.X.X' pattern","selection":{"tagStatus":"tagged","tagPatternList":["v*.*.*"],"countType":"imageCountMoreThan","countNumber":2},"action":{"type":"expire"}},{"rulePriority":2,"description":"Keep max 1 image that is not a tagged version","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1},"action":{"type":"expire"}}]}",
    },
    "RepositoryName": "hour-tracker-members-service",
  },
  "Type": "AWS::ECR::Repository",
  "UpdateReplacePolicy": "Delete",
}
`;

exports[`HourTrackerImageRepositories ORGANIZATION_MANAGER should have the correct repository parameters 1`] = `
{
  "DeletionPolicy": "Delete",
  "Properties": {
    "EmptyOnDelete": true,
    "ImageScanningConfiguration": {
      "ScanOnPush": false,
    },
    "LifecyclePolicy": {
      "LifecyclePolicyText": "{"rules":[{"rulePriority":1,"description":"Keep max 2 versioned images identified by 'vX.X.X' pattern","selection":{"tagStatus":"tagged","tagPatternList":["v*.*.*"],"countType":"imageCountMoreThan","countNumber":2},"action":{"type":"expire"}},{"rulePriority":2,"description":"Keep max 1 image that is not a tagged version","selection":{"tagStatus":"any","countType":"imageCountMoreThan","countNumber":1},"action":{"type":"expire"}}]}",
    },
    "RepositoryName": "hour-tracker-org-manager",
  },
  "Type": "AWS::ECR::Repository",
  "UpdateReplacePolicy": "Delete",
}
`;
